File: Infrastructure/DebugLogger.swift
import Foundation

/// Simple file logger for debugging main functionality
/// Log file is cleared on each app start
class DebugLogger {
    static let shared = DebugLogger()

    private let logFileURL: URL
    private let dateFormatter: DateFormatter
    private let queue = DispatchQueue(label: "com.interviewmaster.logger", qos: .utility)

    private init() {
        // Log file in project directory (use current working directory)
        let cwd = FileManager.default.currentDirectoryPath
        logFileURL = URL(fileURLWithPath: cwd).appendingPathComponent("interview_debug.log")

        dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "HH:mm:ss.SSS"
    }

    /// Clear the log file (call on app start)
    func clear() {
        queue.sync {
            try? "".write(to: logFileURL, atomically: true, encoding: .utf8)
        }
        log("=== Interview Master Started ===")
    }

    /// Log a message with timestamp
    func log(_ message: String, file: String = #file, function: String = #function, line: Int = #line) {
        let timestamp = dateFormatter.string(from: Date())
        let fileName = URL(fileURLWithPath: file).deletingPathExtension().lastPathComponent
        let entry = "[\(timestamp)] [\(fileName):\(line)] \(message)\n"

        queue.async { [weak self] in
            guard let self = self else { return }
            if let handle = try? FileHandle(forWritingTo: self.logFileURL) {
                handle.seekToEndOfFile()
                if let data = entry.data(using: .utf8) {
                    handle.write(data)
                }
                handle.closeFile()
            } else {
                // File doesn't exist, create it
                try? entry.write(to: self.logFileURL, atomically: true, encoding: .utf8)
            }
        }

        // Also print to console for immediate feedback
        NSLog("📝 %@", message)
    }

    /// Log with category prefix
    func log(_ category: LogCategory, _ message: String, file: String = #file, function: String = #function, line: Int = #line) {
        log("\(category.rawValue) \(message)", file: file, function: function, line: line)
    }

    enum LogCategory: String {
        case audio = "🎤 AUDIO:"
        case transcription = "📝 TRANSCRIBE:"
        case classification = "🏷️ CLASSIFY:"
        case answer = "💬 ANSWER:"
        case stream = "📡 STREAM:"
        case error = "❌ ERROR:"
        case delegate = "📤 DELEGATE:"
    }
}

// Convenience global function
func debugLog(_ message: String, file: String = #file, function: String = #function, line: Int = #line) {
    DebugLogger.shared.log(message, file: file, function: function, line: line)
}

func debugLog(_ category: DebugLogger.LogCategory, _ message: String, file: String = #file, function: String = #function, line: Int = #line) {
    DebugLogger.shared.log(category, message, file: file, function: function, line: line)
}

File: Infrastructure/DebugLogger.cs
using System;
using System.IO;
using System.Runtime.CompilerServices;
using System.Threading;

namespace InterviewMasterApp.Infrastructure
{
    /// <summary>
    /// Simple file logger for debugging main functionality
    /// Log file is cleared on each app start
    /// </summary>
    public sealed class DebugLogger
    {
        private static readonly Lazy<DebugLogger> _shared = new(() => new DebugLogger());
        public static DebugLogger Shared => _shared.Value;

        private readonly string _logFilePath;
        private readonly string _timeFormat = "HH:mm:ss.fff";
        private readonly object _fileLock = new();

        private DebugLogger()
        {
            var cwd = Directory.GetCurrentDirectory();
            _logFilePath = Path.Combine(cwd, "interview_debug.log");
        }

        /// <summary>
        /// Clear the log file (call on app start)
        /// </summary>
        public void Clear()
        {
            lock (_fileLock)
            {
                try
                {
                    File.WriteAllText(_logFilePath, string.Empty);
                }
                catch
                {
                    // ignore
                }
            }

            Log("=== Interview Master Started ===");
        }

        /// <summary>
        /// Log a message with timestamp and caller info
        /// </summary>
        public void Log(string message, [CallerFilePath] string file = "", [CallerMemberName] string function = "", [CallerLineNumber] int line = 0)
        {
            var timestamp = DateTime.Now.ToString(_timeFormat);
            var fileName = string.Empty;
            try
            {
                fileName = Path.GetFileNameWithoutExtension(file);
            }
            catch
            {
                fileName = file;
            }

            var entry = $"[{timestamp}] [{fileName}:{line}] {message}{Environment.NewLine}";

            // Async-safe append using a lock to preserve ordering
            lock (_fileLock)
            {
                try
                {
                    File.AppendAllText(_logFilePath, entry);
                }
                catch
                {
                    // If append fails, try to create the file
                    try
                    {
                        File.WriteAllText(_logFilePath, entry);
                    }
                    catch
                    {
                        // swallow
                    }
                }
            }

            // Also print to console for immediate feedback
            try
            {
                Console.WriteLine($"📝 {message}");
            }
            catch
            {
            }
        }

        /// <summary>
        /// Log with category prefix
        /// </summary>
        public void Log(LogCategory category, string message, [CallerFilePath] string file = "", [CallerMemberName] string function = "", [CallerLineNumber] int line = 0)
        {
            Log($"{category.GetPrefix()} {message}", file, function, line);
        }

        public enum LogCategory
        {
            Audio,
            Transcription,
            Classification,
            Answer,
            Stream,
            Error,
            Delegate
        }
    }

    public static class DebugLoggerExtensions
    {
        public static string GetPrefix(this DebugLogger.LogCategory cat) => cat switch
        {
            DebugLogger.LogCategory.Audio => "🎤 AUDIO:",
            DebugLogger.LogCategory.Transcription => "📝 TRANSCRIBE:",
            DebugLogger.LogCategory.Classification => "🏷️ CLASSIFY:",
            DebugLogger.LogCategory.Answer => "💬 ANSWER:",
            DebugLogger.LogCategory.Stream => "📡 STREAM:",
            DebugLogger.LogCategory.Error => "❌ ERROR:",
            DebugLogger.LogCategory.Delegate => "📤 DELEGATE:",
            _ => string.Empty,
        };

        // Convenience helpers to mimic the Swift global functions
        public static void DebugLog(string message, [CallerFilePath] string file = "", [CallerMemberName] string function = "", [CallerLineNumber] int line = 0)
        {
            DebugLogger.Shared.Log(message, file, function, line);
        }

        public static void DebugLog(DebugLogger.LogCategory category, string message, [CallerFilePath] string file = "", [CallerMemberName] string function = "", [CallerLineNumber] int line = 0)
        {
            DebugLogger.Shared.Log(category, message, file, function, line);
        }
    }
}

